app/scripts/background.js

```diff
@@ -663,13 +699,39 @@ export function setupController(
         connectionStream: portStream,
       });
     } else {
+      // this is triggered when a new tab is opened, or origin(url) is changed
       if (remotePort.sender && remotePort.sender.tab && remotePort.sender.url) {
         const tabId = remotePort.sender.tab.id;
         const url = new URL(remotePort.sender.url);
         const { origin } = url;
 
+        // store the orgin to corresponding tab so it can provide infor for onActivated listener
+        if (!Object.keys(tabOriginMapping).includes(tabId)) {
+          tabOriginMapping[tabId] = origin;
+        }
+        const connectSitePermissions =
+          controller.permissionController.state.subjects[origin];
+        // when the dapp is not connected, connectSitePermissions is undefined
+        const isConnectedToDapp = connectSitePermissions !== undefined;
+        // when open a new tab, this event will trigger twice, only 2nd time is with dapp loaded
+        const isTabLoaded = remotePort.sender.tab.title !== 'New Tab';
```

Comment:
This is brittle. I think this may be broken by any extension that takes over the new tab page / seems chromium specific. I'm not sure what a good alternative is though as checking origin would have the same issue i think

Reply:
we really shouldn't be injecting the provider on the chrome new tab page in the first place, but also not sure why we do currently / if there's a good solution not to.

Reply:
> this may be broken by any extension that takes over the new tab page
Not really, we only track dapp in our case. 

We also have some legacy work that we are tracking messages on all the pages currently, (https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/background.js#L172).

Open for new suggestions to track but I have been trying for 2 months and this is the best place i can inject to track opening a new tab and fill in dapp url. :(